romancal/pipeline/exposure_pipeline.py

```diff
@@ -73,123 +68,89 @@ def process(self, input):
-            results.append(result)
+        return_lib = True
+        if file_type == "ModelLibrary":
+            lib = input
+        elif file_type == "asn":
+            lib = ModelLibrary(input)
+        else:
+            # for a non-asn non-library input process it as a library
+            lib = ModelLibrary([input])
+            # but return it as a datamodel
+            return_lib = False
+
+        # Flag to track if any of the input models are fully saturated
+        any_saturated = False
+
+        with lib:
+            for model_index, model in enumerate(lib):
+                self.dq_init.suffix = "dq_init"
+                result = self.dq_init.run(model)
+
+                del model
+
+                result = self.saturation.run(result)
+
+                if is_fully_saturated(result):
+                    log.info("All pixels are saturated. Returning a zeroed-out image.")
+                    result = self.create_fully_saturated_zeroed_image(result)
+
+                    # Track that we've seen a fully saturated input
+                    any_saturated = True
+                    log.warning(
+                        "tweakreg will not be run due to a fully saturated input"
+                    )
+                else:
+                    result = self.refpix.run(result)
+                    result = self.linearity.run(result)
+                    result = self.dark_current.run(result)
+                    result = self.rampfit.run(result)
+                    result = self.assign_wcs.run(result)
+
+                    if result.meta.exposure.type == "WFI_IMAGE":
+                        result = self.flatfield.run(result)
+                        result = self.photom.run(result)
+                        result = self.source_catalog.run(result)
+                    else:
+                        log.info("Flat Field step is being SKIPPED")
+                        log.info("Photom step is being SKIPPED")
+                        log.info("Source Detection step is being SKIPPED")
+                        log.info("Tweakreg step is being SKIPPED")
+                        result.meta.cal_step.flat_field = "SKIPPED"
+                        result.meta.cal_step.photom = "SKIPPED"
+                        result.meta.cal_step.source_detection = "SKIPPED"
+
+                if any_saturated:
+                    # the input association contains a fully saturated model
+                    # where source_catalog can't be run which means we
+                    # also can't run tweakreg.
+                    result.meta.cal_step.tweakreg = "SKIPPED"
+                lib.shelve(result, model_index)
 
         # Now that all the exposures are collated, run tweakreg
         # Note: this does not cover the case where the asn mixes imaging and spectral
         #          observations. This should not occur on-prem
-        result = self.tweakreg.run(results)
+        if not any_saturated:
+            self.tweakreg.run(lib)
```

Comment:
I don't like thinking about the fully saturated case because I expect it to never occur or to be so bad, should it occur, that we don't really care about what we do.

But conceptually if I try to imagine a case where one detector is fully saturated but the others are all fine the "right thing to do" would be to run tweakreg using the source catalogs from the other detectors.  The model that gets fit right now is very rigid and would get well constrained even if only a few detectors were available.

i.e., I guess that I would have imagined the flow being more like: tweakreg loops over the association gathering the catalogs.  If it doesn't find one, it emits some warning message.  If it doesn't find _any_, it gives up and skips or something.  Then the pipeline doesn't need to specially track the any_saturated case here.

Reply:
That sounds good to me.

Given that `source_catalog` is skipped for a fully saturated image I think one option (if it sounds good to you) would be to update tweakreg to check if the image is fully saturated, if so, don't check for the source catalog entry and skip that model. Currently tweakreg throws an exception if an input model doesn't have a source catalog:
https://github.com/spacetelescope/romancal/blob/78c7dc41093f8fb76fe049682ce56de3cad4c488/romancal/tweakreg/tweakreg_step.py#L159
so some sort of change to that will be needed (either turn that into a warning and skip that model or check if it's all saturated).

Reply:
Yeah, I certainly don't consider this a regression, just trying to think about what we want.

Maybe best to go with your current approach to avoid having tweakreg know about full saturation, keeping all of the fully saturated logic in one place.  I hope this case just never actually occurs in practice, so it doesn't matter.

FWIW, I did discuss this, and we could not come up with a situation where this would actually occur for a science exposure.